### ISSUE TITLE
Kapua broker should come with websocket jar included

### ISSUE THREAD

**Issue by hekonsek** [eclipse/kapua#265]
Hi

One of our users requested websocket support for Kapua broker. I think this is reasonable request. IMHO we should include activemq-http jar in our distribution by default and provide a way to enable WS endpoint via environment variable in Docker image.

@dejanb @ctron What do you think, guys?

Cheers!

**Comment by Coduz** [comment]
I'll add @riccardomodanese also in the discussion

**Comment by hekonsek** [comment]
@dejanb is it possible to use ENV variables to enable/disable transport in standalone ActiveMQ distribution? It is certainly possible in embedded ActiveMQ, where you just can add `if` statement :) .

**Comment by sheaphillips** [comment]
@hekonsek is the Dockerfile used to create the kapua broker image available somewhere?  We are aiming to deploy Kapua in OpenShift and hoping to enable websockets in ActiveMQ as we don't want to open ports other than 80 and 443.  One thought was to extend the kapua-broker Dockerfile to add the additional ActiveMQ dependencies for websockets. We'd rather no create a kapua-broker Docker distro from scratch...would rather build on what you've done and contribute back...

**Comment by ccustine** [comment]
@sheaphillips The Dockerfiles are mostly generated from the maven pom.xml in the assembly subproject. There are images for the console, broker, api, etc. For example, here is the console: https://github.com/eclipse/kapua/blob/develop/assembly/pom.xml#L481-L503

You can also find some examples of running on OpenShift here: https://github.com/eclipse/kapua/tree/develop/dev-tools/src/main/openshift

Some of us are working out minor issues with this at the moment but it is nearly 100% there.

**Comment by sheaphillips** [comment]
@ccustine thanks!  that's really helpful.   I think I pretty much see how things are pieced together - may be a little tricky to figure out all the transient deps for activemq-http...

On the openshift side, we've been using the scripts provided - also very helpful!  Our hangup is mainly getting websockets enabled for the broker at this stage.

**Comment by jamesfalkner** [comment]
I created a derived dockerfile from kapua and added in the websockets dependencies (it's way more than the activemq-http.jar file!). I'll post it here in about an hour when I return to my desk..

**Comment by sheaphillips** [comment]
@jamesfalkner terrific! I'm sure that'll save us a lot of futzing around.

**Comment by jamesfalkner** [comment]
As promised, here is what I'm using. There are probably extra files I'm including but I haven't had time to do a real thorough minimization. But it works and I'm able to connect to the broker with a STOMP websockets client like [this one](https://github.com/jmesnil/stomp-websocket).

```
FROM redhatiot/kapua-broker

ENV ACTIVEMQ_TMP /tmp

RUN curl -sSL http://archive.apache.org/dist/activemq/5.14.0/apache-activemq-5.14.0-bin.tar.gz | \
	tar -C /tmp -xvzf - && \
	cp /tmp/apache-activemq-5.14.0/lib/activemq-web-5.14.0.jar /maven/lib && \
	cp -r /tmp/apache-activemq-5.14.0/lib/web /maven/lib && \
	cp /tmp/apache-activemq-5.14.0/lib/optional/*.jar /maven/lib/optional && \
	cp -r /tmp/apache-activemq-5.14.0/webapps /maven && \
	rm -f /maven/lib/optional/guava-12.0.jar && \
	chmod -R a+r /maven/lib/*jar /maven/lib/web/* /maven/lib/optional/*jar /maven/webapps /maven/lib/extra/*jar && \
	rm -rf /tmp/apache-activemq-5.14.0

RUN sed -i "s:</transportConnectors>:\t<transportConnector name=\"websocket\" uri=\"ws\://0.0.0.0\:61614\"/>\n\t</transportConnectors>:" /maven/conf/activemq.xml && sed -i "s:</beans>:\t<import resource=\"jetty.xml\"/>\n\t</beans>:" /maven/conf/activemq.xml

EXPOSE 61614
```

This sets `ACTIVEMQ_TMP` to `/tmp/` because the default value is unwritable in some cases (e.g. on OpenShift).

It then takes a bunch of files from Active MQ 5.14.0 distro and puts them in their right places to add the websockets support. It then adds the transport connector AND un-comments out jetty (which is responsible for handling the websockets web traffic).

**Comment by sheaphillips** [comment]
thanks, @jamesfalkner . I'll give it a spin and let you know how we make out.

**Comment by dejanb** [comment]
@hekonsek It's not, but we can make start it by default (and people can decide if they want to expose it or not)

**Comment by hekonsek** [comment]
@sheaphillips Great to see your interest in Kapua on OpenShift. Ping us if you need anything else :) . As Chris mentioned, we use Fabric8 Docker Maven plugin [1]. It is my personal tool of choice when it comes to generating and publishing Docker images. 

[1] https://github.com/eclipse/kapua/blob/develop/assembly/pom.xml#L457

**Comment by hekonsek** [comment]
@dejanb Maybe we could consider migrating broker to embedded ActiveMQ? It will be possible to enable/disable some broker endpoints and tune behavior via ENV variables? What do you think?

**Comment by sheaphillips** [comment]
@hekonsek thanks!  we appreciate the starting point for OpenShift you've provided. 

We won't be shy about sharing our experiences/suggestions/questions as we get Kapua up and running in our OCP environment.  We'd be happy to contribute back resources such as templates, buildconfigs, deployment configs once we have a better understanding of how we should piece things together.

**Comment by sheaphillips** [comment]
@hekonsek @dejanb re: controlling transport via ENV variables...I personally like the idea of having control over things at runtime vs. having to rebuild an image with different things enabled.  Whether it's ENV variables, or a config injection mechanism (e.g. ConfigMap-provided file in OpenShift to overlay on the default activemq.xml, or some "includes" mechanism for config file), or a combination of strategies... It depends how much variability in configuration is likely to be needed.  If the intention is that the broker is somewhat of a black box within the Kapua stack, then perhaps the amount of configurability is fairly small.  Regardless, being able to turn things on/off without rebuilding depends on the superset of jars required for the "all endpoints" permutation being available in the image.

**Comment by dejanb** [comment]
@hekonsek @sheaphillips It's doable, something like that is done in xPaas. All I was thinking is that there's no reason not to have ws transport started by default. That's the simplest and fastest solution.

**Comment by ctron** [comment]
I would like to adapt the Kura simulator as well to support websockets. As this seems to be the easiest solution for OpenShift. At least for a quick demo.

**Comment by dejanb** [comment]
I'll create a PR today that starts the broker with web socket transport

**Comment by hekonsek** [comment]
OK, so let's add web sockets endpoint by default. We can consider adding ENV-like configuration to the broker in the next version of Kapua, during Artemis migration. Start with a simple solution now and make it more fancy during Artemis migration. What do you think, guys?

**Comment by dejanb** [comment]
@hekonsek Yeah, here's the PR for that
https://github.com/eclipse/kapua/pull/286
I'll modify openshift script

**Comment by dejanb** [comment]
It should be exposed now in Openshift (in ws branch). More testing is needed.

**Comment by ctron** [comment]
:fireworks: Ok .. this is great. I can connect the simulator now using `ws://cluster-ip:80`!

**Comment by ctron** [comment]
Ok .. I may have been a bit too enthusiastic about this one. I get a connection using websockets but the broker throws the following exception:

```
13:45:15,803 [Camel (camelContext) thread #2 - JmsConsumer[Consumer.eurotech:mainRoutelisteners:EXACTLY_ONCE.VirtualTopic.%3E]] WARN  ErrorMessageListener:86  - Processing error message for service generic... Message type [B - Endpoint null - Error message null
13:45:15,803 [Camel (camelContext) thread #2 - JmsConsumer[Consumer.eurotech:mainRoutelisteners:EXACTLY_ONCE.VirtualTopic.%3E]] WARN  ErrorMessageListener:92  - Exception: 
java.lang.NullPointerException
	at org.eclipse.kapua.broker.core.message.JmsUtil.convertToCamelKapuaMessage(JmsUtil.java:159)
	at org.eclipse.kapua.broker.core.converter.AbstractKapuaConverter.convertTo(AbstractKapuaConverter.java:85)
	at org.eclipse.kapua.broker.core.converter.KapuaLifeCycleConverter.convertToBirth(KapuaLifeCycleConverter.java:78)
	at sun.reflect.GeneratedMethodAccessor49.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.camel.component.bean.MethodInfo.invoke(MethodInfo.java:408)
	at org.apache.camel.component.bean.MethodInfo$1.doProceed(MethodInfo.java:279)
	at org.apache.camel.component.bean.MethodInfo$1.proceed(MethodInfo.java:252)
	at org.apache.camel.component.bean.BeanProcessor.process(BeanProcessor.java:177)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:460)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:121)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:83)
	at org.apache.camel.processor.ChoiceProcessor.process(ChoiceProcessor.java:117)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:460)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.ChoiceProcessor.process(ChoiceProcessor.java:117)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:460)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:121)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:83)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:460)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:190)
	at org.apache.camel.util.AsyncProcessorHelper.process(AsyncProcessorHelper.java:109)
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:87)
	at org.apache.camel.component.jms.EndpointMessageListener.onMessage(EndpointMessageListener.java:112)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:689)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:649)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:619)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:307)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:245)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1144)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1136)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1033)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by ctron** [comment]
I think the cause for this is here: https://github.com/eclipse/kapua/blob/5fbbca64359dd2042bbc7c058cd4b1cf3a2e54d1/broker-core/src/main/java/org/eclipse/kapua/broker/core/plugin/ConnectorDescriptorLoader.java#L67

**Comment by ctron** [comment]
Although PR #290 fixes the NPE and lets the broker create device entries, they are all marked as "disconnected".

**Comment by ctron** [comment]
Looking at the table `DVC_DEVICE` I can see that all WebSocket based connections have an empty `CONNECTION_ID` column. Not sure why though.

**Comment by ctron** [comment]
Ok ... PEBKAC ... I was using `kapua-sys` instead of `kapua-broker` as user for the broker. However this seems to completely screw up the internal state of Kapua. But I guess we should track this under a different issue.

Still PR #290 is required for the Websocket support.

**Comment by hekonsek** [comment]
Take a look at this - #314. I think that we should try to get rid of this predefined list of ConnectorDesciptors.

**Comment by dejanb** [comment]
We should create a test for this, once we make it working

**Comment by mcarrer** [comment]
Closed by PR #290
